kpi/models/object_permission.py

```diff
@@ -948,3 +962,194 @@ def _update_partial_permissions(self, user_id, perm, remove=False,
+                    ...
+                ],
+                '<object_id>': [
+                    (permission_id, permission_codename, deny),
+                    (permission_id, permission_codename, deny),
+                    ...
+                ]
+            }
+        """
+        records = ObjectPermission.objects. \
+            filter(content_type_id=content_type_id, user=user_id).\
+            values('object_id',
+                   'permission_id',
+                   'permission__codename',
+                   'deny')
+
+        object_permissions_per_object = defaultdict(list)
+        for record in records:
+            object_permissions_per_object[record['object_id']].append((
+                record['permission_id'],
+                record['permission__codename'],
+                record['deny'],
+            ))
+
+        return object_permissions_per_object
+
+    def __get_object_permissions(self, is_denied, user=None, codename=None):
+        """
+        Returns a set of user ids and object permission ids related to
+        object `self`.
+
+        Args:
+            is_denied (bool): If `True`, returns denied permissions
+            user (User)
+            codename (str)
+
+        Returns:
+            set: [(User's pk, Permission's pk)]
+        """
+
+        def build_dict(user_id_, object_permissions_):
+            perms_ = []
+            if object_permissions_:
+                for permission_id, codename_, deny in object_permissions_:
+                    if (deny is not is_denied or
+                            codename is not None and
+                            codename != codename_):
+                        continue
+                    perms_.append((user_id_, permission_id))
+            return perms_
+
+        perms = []
+        
+        # If User is not none, retrieve all permissions for this user
+        # grouped by object ids, otherwise, retrieve all permissions for this object
+        # grouped by user ids.
+        if user is not None:
+            user_id = user.pk if not user.is_anonymous() \
+                else settings.ANONYMOUS_USER_ID
+            all_object_permissions = self.__get_all_user_permissions(
+                content_type_id=ContentType.objects.get_for_model(self).pk,
+                user_id=user_id)
+            perms = build_dict(user_id, all_object_permissions.get(self.pk))
+        else:
+            all_object_permissions = self.__get_all_object_permissions(
+                content_type_id=ContentType.objects.get_for_model(self).pk,
+                object_id=self.pk)
+            for user_id, object_permissions in all_object_permissions.items():
+                perms += build_dict(user_id, object_permissions)
+
+        return set(perms)
+
+    @staticmethod
+    @cache_for_request
+    def __get_permissions_for_content_type(content_type_id, codenames,
+                                           startswith=True, first=False):
```

Comment:
tbh, writing `__get_permissions_for_content_type(…, first=True)`, or even `__get_permissions_for_content_type(…, True)`, when calling this method is longer than typing `__get_permissions_for_content_type(…)[0]` :wink: